File: LeetCode/Medium/0216-combination-sum-iii/0216-combination-sum-iii.cpp
class Solution {
public:
    vector<vector<int>> ret;
    int k;
    int n;
    int last;
    
    vector<vector<int>> combinationSum3(int k, int n) {
        this->k = k;
        this->n = n;
        last = min(n, 10);
        
        vector<int> temp(k, 0);
        DFS(temp, 0, 0, 1);
        
        return ret;
    }
    
    void DFS(vector<int> curr, int idx, int currSum, int start){
        if (idx == k){
            if (currSum == n) ret.push_back(curr);
            return;
        }
        
        for (int i = start; i < last; ++i){
            if (currSum + i > n) return;
            curr[idx] = i;
            DFS(curr, idx + 1, currSum + i, i + 1);
        }
    }
};

File: LeetCode/Medium/0216-combination-sum-iii/0216-combination-sum-iii.go
func combinationSum3(k int, n int) [][]int {
    ret := make([][]int, 0)
    last := 10
    if n < last { last = n }
    
    dfs := func(curr []int, idx int, start int, currSum int){}
    dfs = func(curr []int, idx int, start int, currSum int){
        if idx == k {
            if currSum == n { ret = append(ret, curr) }
            return
        }
        
        for i := start; i < last; i++ {
            if currSum + i > n { return }
            curr[idx] = i
            dfs(append([]int(nil), curr...), idx + 1, i + 1, currSum + i)
        }
    }
    
    dfs(make([]int, k), 0, 1, 0)
    return ret
}